dsp/dsp.py

```diff
@@ -0,0 +1,254 @@
+        window = np.ones(window_size)
+        return window, window_size
+    window_size = window_size - 1
+    if window_type == 'bartlett':
+        window = sig.windows.bartlett(window_size)
+        overlap_factor = 0.5
+    elif window_type == 'hann':
+        window = sig.windows.hann(window_size)
+        overlap_factor = 0.5
+    elif window_type == 'blackman':
+        window = sig.windows.blackman(window_size)
+        overlap_factor = 2.0 / 3.0
+    elif window_type == 'blackmanharris':
+        window = sig.windows.blackmanharris(window_size)
+        overlap_factor = 0.75
+    else:
+        raise Exception(f"Unsupported window type {window_type}. Must be one of {', '.join(window_types)}")
+    
+    assert(overlap_factor >= 0.0 and overlap_factor <= 1.0)
+    
+    hop_factor = 1.0 - overlap_factor
+    initial_hop_size = int(np.round(hop_factor * window_size))
+
+    # COLA check
+    best_hop_size = initial_hop_size
+    best_gain_factor = 1.0
+    best_error = np.Infinity
+    for hop_size in [initial_hop_size, initial_hop_size - 1]:
+        num_windows = int(2.0 / hop_factor)
+        cola_check = np.zeros(num_windows * hop_size + (window_size - hop_size))
+        for i in range(num_windows):
+            n_start = i * hop_size
+            n_end = n_start + window_size
+            cola_check[n_start : n_end] += window
+        gain_factor = 1.0 / np.max(cola_check)
+        cola_check *= gain_factor
+
+        cola_start = window_size - hop_size // 2
+        cola_end = cola_start + window_size - hop_size
+        cola_check = cola_check[cola_start : cola_end]
+        cola_check /= np.max(cola_check)
+
+        error = np.sum(np.ones_like(cola_check) - cola_check)
+        if error < best_error:
+            best_error = error
+            best_hop_size = hop_size
+            best_gain_factor = gain_factor
+    
+    hop_size = best_hop_size
+    gain_factor = best_gain_factor
+    window = np.append(window * gain_factor, np.zeros(1))
+
+    return window, hop_size
+
+def get_windowed_frame_size(config):
+    window_size = config['window_size']
+    padding_factor = config['padding_factor']
+    frame_size = window_size * pow(2, padding_factor)
+    return frame_size
+
+def windowed_frame_analysis(signal, config):
+    window = config['window']
+    window_size = config['window_size']
+    assert(len(window) == window_size)
+    hop_size = config['hop_size']
+    assert(hop_size <= window_size)
+    padding_factor = config['padding_factor']
+    assert(int(padding_factor) == padding_factor)
+    
+    zero_phase = config['zero_phase']
+    sqrt_window = config['sqrt_window']
+
+    N = len(signal)
+    H = hop_size
+    num_hops = (N - window_size) // H + 1
+    M = num_hops * H + window_size
+    if M > N:
+        signal = np.append(signal, np.zeros(M - N))
+    assert(len(signal) == M)
+    
+    num_hops = (M - window_size) // H + 1
+    
+    if sqrt_window and np.all(window > 0):
+        window = np.sqrt(window)
```

Comment:
Shouldn't this raise an exception when sqrt_window is true but some values are negative?

Reply:
yes it should probably raise an exception instead of failing silently, I'll add that